neqo-common/src/codec.rs

```diff
@@ -197,12 +202,120 @@ impl<'b> PartialEq<Decoder<'b>> for Decoder<'_> {
 }
 
 /// Encoder is good for building data structures.
-#[derive(Clone, Default, PartialEq, Eq)]
-pub struct Encoder {
-    buf: Vec<u8>,
+#[derive(Clone, PartialEq, Eq)]
+pub struct Encoder<B = Vec<u8>> {
+    buf: B,
+}
+
+impl Default for Encoder {
+    fn default() -> Self {
+        Self { buf: Vec::new() }
+    }
+}
+
+impl<B: std::io::Write> Encoder<B> {
+    /// Generic encode routine for arbitrary data.
+    ///
+    /// # Panics
+    ///
+    /// When writing to the underlying buffer fails.
+    pub fn encode(&mut self, data: &[u8]) -> &mut Self {
+        self.buf.write_all(data).expect("TODO");
+        self
+    }
+
+    /// Encode a single byte.
+    ///
+    /// # Panics
+    ///
+    /// When writing to the underlying buffer fails.
+    pub fn encode_byte(&mut self, data: u8) -> &mut Self {
+        self.buf.write_all(&[data]).expect("TODO");
+        self
+    }
+
+    /// Encode an integer of any size up to u64.
+    ///
+    /// # Panics
+    ///
+    /// When `n` is outside the range `1..=8`.
+    pub fn encode_uint<T: Into<u64>>(&mut self, n: usize, v: T) -> &mut Self {
+        let v = v.into();
+        assert!(n > 0 && n <= 8);
+        for i in 0..n {
+            self.encode_byte(((v >> (8 * (n - i - 1))) & 0xff) as u8);
+        }
+        self
+    }
+
+    /// Encode a QUIC varint.
+    ///
+    /// # Panics
+    ///
+    /// When `v >= 1<<62`.
+    pub fn encode_varint<T: Into<u64>>(&mut self, v: T) -> &mut Self {
+        let v = v.into();
+        match () {
+            () if v < (1 << 6) => self.encode_uint(1, v),
+            () if v < (1 << 14) => self.encode_uint(2, v | (1 << 14)),
+            () if v < (1 << 30) => self.encode_uint(4, v | (2 << 30)),
+            () if v < (1 << 62) => self.encode_uint(8, v | (3 << 62)),
+            () => panic!("Varint value too large"),
+        };
+        self
+    }
+
+    /// Encode a vector in TLS style.
+    ///
+    /// # Panics
+    ///
+    /// When `v` is longer than 2^64.
```

Comment:
```suggestion
    /// When `v` is longer than 2^`n`.
```

Reply:
Simply copied from previous implementation. Will fix in a separate pull request.

Reply:
Given this simply being copied from `main`, I propose we fix this in a separate pull request.

https://github.com/mozilla/neqo/pull/2735